src/server/lib/okta/openid-connect.ts

```diff
@@ -0,0 +1,252 @@
+ */
+export const ProfileOpenIdClientRedirectUris: OpenIdClientRedirectUris = {
+  WEB: `${getProfileUrl()}/oauth/authorization-code/callback`,
+};
+
+/**
+ * A subset of the 'openid-client' lib `Client` class
+ * features/mechanisms to expose.
+ * The client/app is the "profile" OAuth app in Okta
+ * @property `authorizationUrl` - Generate the `/authorize` url for the Authorization Code Flow (w or w/o PKCE)
+ * @property `callbackParams` - Get OpenID Connect query parameters returned to the callback (redirect_uri)
+ * @property `oauthCallback` - Method used in the callback (redirect_uri) endpoint to get OAuth tokens
+ */
+export const ProfileOpenIdClient = new OIDCIssuer.Client({
+  client_id: okta.clientId,
+  client_secret: okta.clientSecret,
+  redirect_uris: Object.values(ProfileOpenIdClientRedirectUris),
+}) as OpenIdClient;
+
+/**
+ * Possible `error` types return by OpenID Connect and social login flows with okta
+ * Sourced from https://developer.okta.com/docs/reference/error-codes/#example-errors-for-openid-connect-and-social-login
+ */
+export enum OpenIdErrors {
+  UNAUTHORIZED_CLIENT = 'unauthorized_client',
+  ACCESS_DENIED = 'access_denied',
+  UNSUPPORTED_RESPONSE_TYPE = 'unsupported_response_type',
+  INVALID_SCOPE = 'invalid_scope',
+  SERVER_ERROR = 'server_error',
+  TEMPORARILY_UNAVAILABLE = 'temporarily_unavailable',
+  INVALID_CLIENT = 'invalid_client',
+  LOGIN_REQUIRED = 'login_required',
+  INVALID_REQUEST = 'invalid_request',
+  USER_CANCELED_REQUEST = 'user_canceled_request',
+}
+
+/**
+ * Generate a cryptographically secure random string to be used
+ * as the `state` parameter in the authorization code flow
+ * which is used as the state parameter value to prevent CSRF attacks.
+ * @param {number} [bytes=16]
+ * @return {string}
+ */
+const generateRandomString = (bytes = 16): string =>
+  randomBytes(bytes).toString('hex');
+
+/**
+ * Type guard to check that the parsed cookie is actually
+ * of type Authorization state
+ *
+ * @param {unknown} obj
+ * @return {boolean}
+ */
+const isAuthorizationState = (obj: unknown): obj is AuthorizationState => {
+  const objAsAuthState = obj as AuthorizationState;
+  return !!(objAsAuthState.stateParam && objAsAuthState.queryParams);
+};
+
+/**
+ * Generate the `AuthorizationState` object to be stored as a
+ * cookie on the client, and the `stateParam` used by the Authorization
+ * Code flow.
+ * @param {string} returnUrl
+ * @return {*} `AuthorizationState`
+ */
+export const generateAuthorizationState = (
+  queryParams: PersistableQueryParams,
+): AuthorizationState => ({
+  stateParam: generateRandomString(),
+  queryParams,
+});
+
+/**
+ * Name of the Authorization State cookie to set
+ * @type {string}
+ */
+const AuthorizationStateCookieName = 'GU_oidc_auth_state';
+
+/**
+ * Authorization state cookie options
+ * No expiry/max-age, so cookie is deleted when browser session closes
+ */
+const AuthorizationStateCookieOptions: CookieOptions = {
+  httpOnly: true,
```

Comment:
In `encryptedStateCookieOptions`, these boolean values are set using: `!baseUri.includes('localhost')`, might be completely unnecessary but just wanted to double check that this doesn't need the same

Reply:
ah good shout, I'll update the options to use the localhost check, otherwise things won't work on that domain as expected